### ISSUE TITLE
journald - Consider parsing more known fields from logs

### ISSUE THREAD

**Issue by djaglowski** [open-telemetry/opentelemetry-collector-contrib#7298]
...

I'm looking at the journald input, which currently dumps everything into the body.  The only thing it pulls out is the timestamp.

Ideally (IMHO), the body would be a string containing MESSAGE and all the other fields stored as attributes.  Additionally, the priority would be converted to an OpenTelemetry severity.  Given journald has a well defined list of fields [1], many of them could be converted to semantic conventions such as `process.pid`.

So, the question is whether things like the journald input should convert messages to a more native opentelemetry format?  If not, should there be an option, operator or processor that could do it all in one go?

(I can create a PR to change the journald input if there is agreement)

1. https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html

_Originally posted by @gregoryfranklin in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3129#issuecomment-1017782954_

**Comment by dmitryax** [comment]
I would like to see journald logs to be parsed right OOTB. I don't know why it shouldn't be done right at the source without any additional operator or processor. Maybe we can provide an option to keep the raw message instead of parsing if user wants that?

**Comment by djaglowski** [comment]
I support that the operator ought to do further parsing as well. We already do this in `syslog_input` to an extent. Some users may prefer the collector to do less parsing, but unless someone is asking for it, I think it's reasonable to add "don't parse" controls later as necessary.

Mapping from priority to severity, the body being only the MESSAGE string, and everything else going to attributes makes sense to me. I also agree with applying semantic conventions as they currently apply. In the long term, I assume we'll have semantic conventions for most/all the fields, and I would personally like to see this operator apply them wherever possible. 

What's less clear to me is whether we should work ahead of the semantic conventions. Given that the library is not yet stable and that development can easily outpace semantic debate, I think it's probably reasonable to make a best effort at appropriately naming and then propose the attributes to the spec repo.

What thoughts do others have on all this?

**Comment by dmitryax** [comment]
I fully agree on the first two paragraphs.

> Given that the library is not yet stable and that development can easily outpace semantic debate, I think it's probably reasonable to make a best effort at appropriately naming and then propose the attributes to the spec repo.

I see your point, but from the other hand we can cause additional breaking changes to the attributes that the library emit if our best effort doesn't survive the semantic debate. I would go with semantic conventions first, or in parallel, to make sure that once the change to the library is released at least attribute names are not going to change.

**Comment by gregoryfranklin** [comment]
I'll start with a PR that moves all the fields to attributes instead of body and sets the severity.

So, it will change from this
```
"entry": {
  "timestamp": "2020-04-16T11:05:49.516168-04:00",
  "body": {
    "CODE_FILE": "../src/core/unit.c",
    "CODE_FUNC": "unit_log_success",
    "CODE_LINE": "5487",
    "MESSAGE": "var-lib-docker-overlay2-bff8130ef3f66eeb81ce2102f1ac34cfa7a10fcbd1b8ae27c6c5a1543f64ddb7-merged.mount: Succeeded.",
    "MESSAGE_ID": "7ad2d189f7e94e70a38c781354912448",
    "PRIORITY": "6",
```
to this
```
"entry": {
  "timestamp": "2020-04-16T11:05:49.516168-04:00",
  "body": "var-lib-docker-overlay2-bff8130ef3f66eeb81ce2102f1ac34cfa7a10fcbd1b8ae27c6c5a1543f64ddb7-merged.mount: Succeeded.",
  "severity": 9,
  "severity_text", "INFO",
  "attributes": {
    "CODE_FILE": "../src/core/unit.c",
    "CODE_FUNC": "unit_log_success",
    "CODE_LINE": "5487",
    "MESSAGE_ID": "7ad2d189f7e94e70a38c781354912448",
```

Some of the fields are also clearly resource attributes rather than log attributes, but I think that is linked to the semantic conventions discussion so will leave for a later PR.

I agree it makes sense to wait for the discussion on semantic conventions to move further along before applying these to the library.  In the spec there is an example mapping for syslog (and many other common logging formats) - it would be nice to see the same for journald.

**Comment by jsirianni** [comment]
Has there been any movement on this? If not, I would be happy to jump in. I have been working on integrating Journald into a few projects and it has been frustrating trying to move all fields over to attributes.

**Comment by djaglowski** [comment]
I don't see any reason why this shouldn't move forward. Putting this info in attributes is clearly an improvement in terms of respecting the data model.

**Comment by gregoryfranklin** [comment]
I had a go at this a while back but the decision was to open an issue against the specification to get the missing semantic conventions defined.
You'll see in the PR (https://github.com/open-telemetry/opentelemetry-log-collection/pull/353) that I identified all the well-defined fields from journald and which of them could be mapped to existing semantic conventions.

But, I've not had time to follow up on it in terms of opening an issue against the spec or anything.

**Comment by djaglowski** [comment]
There is an [ongoing discussion](https://github.com/open-telemetry/opentelemetry-specification/issues/2610#issuecomment-1190784629) about whether and how semantic conventions should be defined on a per component basis, so it is not even fully settled that these attributes will ever be codified. At this point, I don't think we should be waiting on semantic conventions. Many other components are moving forward with best effort attribute naming. This is an alpha components anyways, so names are not irreversible.

What we know for sure at this point is that this integration is not adhering to the [data model's direction that first-party libraries should emit logs with the body as a string](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-body). We should correct this behavior. 

Since adhering to the data model will require us to set attributes, we could either 1) follow the style and spirit of the semantic conventions, making a best effort to establish reasonable attribute names, or 2) use the raw values from journald (eg. `_MACHINE_ID`). The semantic conventions provide detailed guidance on [naming attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md#attribute-naming), so I believe we should apply these as best we can to this domain.

**Comment by djaglowski** [comment]
This is still relevant, but should wait until a final decision on the [ECS OTEP](https://github.com/open-telemetry/oteps/pull/199).

**Comment by cwegener** [comment]
> This is still relevant, but should wait until a final decision on the [ECS OTEP](https://github.com/open-telemetry/oteps/pull/199).

Since ECS OTEP has been merged, does that change anything regarding this journald issue? I can't see anything about the ECS OTEP that would have an impact.

**Comment by fionera** [comment]
I am also interested in this. Should I revive the PR that @gregoryfranklin made?

**Comment by cwegener** [comment]
Was there a Google Doc already to work on Journald semantics?

**Comment by djaglowski** [comment]
> Was there a Google Doc already to work on Journald semantics?

I'm not aware of one.

**Comment by cwegener** [comment]
> Hi, I'm interested in this too :) @djaglowski @cwegener is there any work being done that it's not tracked? Anything that can be done to help out?

I haven't done any work on this. I think the first step would be to review the fields in Journald and figure out how many of them should/can be mapped into OTel Log attributes. And if there is any impact on existing or not yet existing semantic conventions.

The available Journald fields are: https://www.freedesktop.org/software/systemd/man/latest/systemd.journal-fields.html

**Comment by FeltrinN** [comment]
Sounds good. I'll try putting together a table unless somebody objects :) Once I've got a reasonable draft I'll open a documentation PR...

**Comment by FeltrinN** [comment]
>  (e.g. I can't tell if the way how the severity mapping was done in the old PR would still how it should be done today)

I've done some very [early stages work](https://github.com/FeltrinN/opentelemetry-collector-contrib/blob/7298-journald-field-docs/receiver/journaldreceiver/FieldMappings.md), but I think it's clear already that the severity mapping from the earlier PR is valid (and quite stable, as those mappings are already in an appendix to the semantic conventions). Therefore, if anyone feels like it, I think it should be okay to split the old PR and merge the severity mappings already, as getting the fields in order will probably take a bit longer...

**Comment by yaleman** [comment]
On this topic, is there an easy-ish way of doing this with transforms/existing tooling today?

**Comment by t-davies** [comment]
> On this topic, is there an easy-ish way of doing this with transforms/existing tooling today?

This should get you most of the way there, 100% behind having this done automatically though.

```(yaml)
processors:
  transform:
    log_statements:
      - context: log
        statements:
          - set(severity_number, SEVERITY_NUMBER_DEBUG) where body["PRIORITY"] == 7
          - set(severity_number, SEVERITY_NUMBER_INFO) where body["PRIORITY"] == 6
          - set(severity_number, SEVERITY_NUMBER_INFO2) where body["PRIORITY"] == 5
          - set(severity_number, SEVERITY_NUMBER_WARN) where body["PRIORITY"] == 4
          - set(severity_number, SEVERITY_NUMBER_ERROR) where body["PRIORITY"] == 3
          - set(severity_number, SEVERITY_NUMBER_FATAL) where body["PRIORITY"] >= 2
          - set(attributes["process.command_line"], body["_CMDLINE"])
          - set(attributes["process.command"], body["_COMM"])
          - set(attributes["process.executable.path"], body["_EXE"])
          - set(body, body["MESSAGE"])
 ```

**Comment by yaleman** [comment]
Thanks @t-davies that works great.

**Comment by steverweber** [comment]
regarding t-Davies solution.

Should this be set on "attributes" instead of "resource" ?

```
          - set(attributes["process.pid"], body["_PID"])
```

I'm wondering if "attributes" is more so geared for high cardinality and important for some database backends to prevent sharding like splitting. If this is true it would be nice to have this touched on in docs https://opentelemetry.io/docs/concepts/resources/

**Comment by snv** [comment]
Hi,
I found this thread while looking why so much information gets lost along the way from my logs.

I just want to add, that my use case is, that i see in grafana only
![image](https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/992077/acedccc9-5208-4940-b5c7-8d0ba343bc96)

while the debug output from the collector shows that the input was a lot richer

> 2024-06-24T07:27:50.198+0200    info    ResourceLog #0
Resource SchemaURL: https://opentelemetry.io/schemas/1.6.1
Resource attributes:
     -> host.name: Str(vs-devop-tempo-01.optadata-gruppe.local)
     -> os.type: Str(linux)
ScopeLogs #0
ScopeLogs SchemaURL: 
InstrumentationScope  
LogRecord #0
ObservedTimestamp: 2024-06-24 05:27:50.008482524 +0000 UTC
Timestamp: 2024-06-24 05:27:49.533141 +0000 UTC
SeverityText: 
SeverityNumber: Unspecified(0)
Body: Map({"MESSAGE":"level=info ts=2024-06-24T05:27:49.533082034Z caller=registry.go:236 tenant=single-tenant msg=\"collecting metrics\" active_series=2091","PRIORITY":"6","SYSLOG_FACILITY":"3","SYSLOG_IDENTIFIER":"tempo","_BOOT_ID":"18d2abf8395d4333b025bea956c1d5d7","_CAP_EFFECTIVE":"0","_CMDLINE":"/usr/bin/tempo -config.file /etc/tempo/config.yml","_COMM":"tempo","_EXE":"/usr/bin/tempo","_GID":"65534","_HOSTNAME":"vs-devop-tempo-01","_MACHINE_ID":"a48d0459d22ae4ec8553c99165c0af6f","_PID":"3249697","_RUNTIME_SCOPE":"system","_SELINUX_CONTEXT":"unconfined\n","_STREAM_ID":"7dfe21fbf3ee44e795a16b017a0d6648","_SYSTEMD_CGROUP":"/system.slice/tempo.service","_SYSTEMD_INVOCATION_ID":"9a5cfe40ac87458c9121038a00d2b049","_SYSTEMD_SLICE":"system.slice","_SYSTEMD_UNIT":"tempo.service","_TRANSPORT":"stdout","_UID":"111","__CURSOR":"s=11409092ba2941e59ab2a6295421fa01;i=77cb61;b=18d2abf8395d4333b025bea956c1d5d7;m=afca6051f9b;t=61b9c0ac101d5;x=5f64b366e0f30616","__MONOTONIC_TIMESTAMP":"12080233389979"})
Trace ID: 
Span ID: 
Flags: 0
        {"kind": "exporter", "data_type": "logs", "name": "debug"}

I really would like to at least be able to filter my logs in grafana by the values from `SYSLOG_IDENTIFIER` or `_SYSTEMD_UNIT`, but those are not converted to attributes, yet.

Can i -- as a workaround -- archive that using the `operators` configuration?

**Comment by cwegener** [comment]
> Can i -- as a workaround -- archive that using the `operators` configuration?

I would recommend using a `transform` processor instead. e.g.

```yaml
  transform/journald:
    error_mode: ignore
    log_statements:
      - context: log
        statements:
          - set(cache, body)
          - set(attributes["systemd.unit"], cache["_SYSTEMD_UNIT"]) where cache["_SYSTEMD_UNIT"] != nil
```

`transform` processors can run on the Gateway collector, whereas `operators` would typically run in the `journald` receiver on the Agent collector instead. Shifting the heavier processing from the Agent to the Gateway is usually preferred.

**Comment by steverweber** [comment]
perhaps it'll help someone, here's what im using. The attributes["job"] might be handy.

``` yaml
receivers:
  journald:
    #priority: debug #(default: info)
    resource: # tag or label the receiver
      receiver: otel-journald
processors:
  transform/journald:
    error_mode: ignore
    log_statements:
      - context: log
        statements:
          # https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md
          #--- severity level priority
          - set(severity_number, SEVERITY_NUMBER_DEBUG) where Int(body["PRIORITY"]) == 7
          - set(severity_number, SEVERITY_NUMBER_INFO) where Int(body["PRIORITY"]) == 6
          - set(severity_number, SEVERITY_NUMBER_INFO2) where Int(body["PRIORITY"]) == 5
          - set(severity_number, SEVERITY_NUMBER_WARN) where Int(body["PRIORITY"]) == 4
          - set(severity_number, SEVERITY_NUMBER_ERROR) where Int(body["PRIORITY"]) == 3
          - set(severity_number, SEVERITY_NUMBER_FATAL) where Int(body["PRIORITY"]) <= 2
          - set(attributes["priority"], body["PRIORITY"])
          #unneeded - set(attributes["level"], severity_number)
          #--- trace who and what generated log message
          - set(attributes["process.comm"], body["_COMM"])
          - set(attributes["process.exec"], body["_EXE"])
          - set(attributes["process.uid"], body["_UID"])
          - set(attributes["process.gid"], body["_GID"])
          - set(attributes["owner_uid"], body["_SYSTEMD_OWNER_UID"])
          - set(attributes["unit"], body["_SYSTEMD_UNIT"])
          - set(attributes["syslog_identifier"], body["SYSLOG_IDENTIFIER"])
          #--- verbose extras
          #- set(attributes["process.pid"], body["_PID"])
          #- set(attributes["process.cmdline"], body["_CMDLINE"])
          #- set(attributes["_SYSTEMD_SLICE"], body["_SYSTEMD_SLICE"])
          #- set(attributes["runtime_scope"], body["_RUNTIME_SCOPE"])
          #--- create low cardinality "job" identifier
          # user@xxx.service mnt-xxx.mount run-xxx.scope session-xxx.scope
          ## ^([a-zA-Z_]{3,20})   ([^\\-\\.\\@0-9]+).*
          - set(attributes["syslog_identifier_prefix"], ConvertCase(body["SYSLOG_IDENTIFIER"], "lower")) where body["SYSLOG_IDENTIFIER"] != nil
          - replace_pattern(attributes["syslog_identifier_prefix"], "^[^a-zA-Z]*([a-zA-Z]{3,25}).*", "$$1") where body["SYSLOG_IDENTIFIER"] != nil
          - set(attributes["unit_prefix"], ConvertCase(body["_SYSTEMD_UNIT"], "lower")) where body["_SYSTEMD_UNIT"] != nil
          - replace_pattern(attributes["unit_prefix"], "^[^a-zA-Z]*([a-zA-Z]{3,25}).*", "$$1") where body["_SYSTEMD_UNIT"] != nil
          - set(attributes["job"], attributes["syslog_identifier_prefix"])
          - set(attributes["job"], attributes["unit_prefix"]) where attributes["job"] == nil and attributes["unit_prefix"] != nil
          #- delete_key(attributes, "syslog_identifier_prefix")
          #- delete_key(attributes, "unit_prefix")
          #--- remove jorunald metadata and make log body simple
          - set(body, body["MESSAGE"])
  groupbyattrs:
    keys:
      - service.name
      - host.name
      - receiver
      - job
service:
  pipelines:
    logs/journald:
      receivers: [journald]
      processors: [transform/journald, batch, resource/default, resourcedetection/system, groupbyattrs]
      exporters: [loki/VictoriaLogs, otlphttp/loki]
```

**Comment by fredleger** [comment]
Faced this one as well.
My use case was to send logs to quickwit (https://github.com/quickwit-oss/quickwit/) from journald.
Quickwit by default will drop what is not an attribute. So i must add @steverweber snippet to see any log (thanks for that btw).

Any update on the progression of it ? Any help needed ?

**Comment by fredleger** [comment]
based on the work of @steverweber i made a version more suitable for my specific needs with quickwit. If that can help someone ...

```yaml
  transform/journald:
    error_mode: ignore
    log_statements:
      - context: log
        statements:
          # https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md
          #--- debug
          # - set(attributes["original_body"], body.string)
          #--- severity level priority
          - set(severity_number, SEVERITY_NUMBER_DEBUG) where Int(body["PRIORITY"]) == 7
          - set(severity_number, SEVERITY_NUMBER_INFO)  where Int(body["PRIORITY"]) == 6
          - set(severity_number, SEVERITY_NUMBER_INFO2) where Int(body["PRIORITY"]) == 5
          - set(severity_number, SEVERITY_NUMBER_WARN)  where Int(body["PRIORITY"]) == 4
          - set(severity_number, SEVERITY_NUMBER_ERROR) where Int(body["PRIORITY"]) == 3
          - set(severity_number, SEVERITY_NUMBER_FATAL) where Int(body["PRIORITY"]) <= 2
          - set(attributes["priority"], body["PRIORITY"])
          #--- severity level priority as text
          - set(attributes["severity"], "debug")  where Int(body["PRIORITY"]) == 7
          - set(attributes["severity"], "info")   where Int(body["PRIORITY"]) == 6
          - set(attributes["severity"], "info")   where Int(body["PRIORITY"]) == 5
          - set(attributes["severity"], "warn")   where Int(body["PRIORITY"]) == 4
          - set(attributes["severity"], "error")  where Int(body["PRIORITY"]) == 3
          - set(attributes["severity"], "fatal")  where Int(body["PRIORITY"]) <= 2
          #--- trace who and what generated log message
          - set(attributes["process.comm"],       String(body["_COMM"]))                where body["_COMM"] != nil
          - set(attributes["process.exec"],       String(body["_EXE"]))                 where body["_EXE"] != nil
          - set(attributes["process.uid"],        String(body["_UID"]))                 where body["_UID"] != nil
          - set(attributes["process.gid"],        String(body["_GID"]))                 where body["_GID"] != nil
          - set(attributes["owner_uid"],          String(body["_SYSTEMD_OWNER_UID"]))   where body["_SYSTEMD_OWNER_UID"] != nil
          - set(attributes["unit"],               String(body["_SYSTEMD_UNIT"]))        where body["_SYSTEMD_UNIT"] != nil
          - set(attributes["syslog_identifier"],  String(body["SYSLOG_IDENTIFIER"]))    where body["SYSLOG_IDENTIFIER"] != nil
          #--- verbose extras
          - set(attributes["process.pid"],      String(body["_PID"]))           where body["_PID"] != nil
          - set(attributes["process.cmdline"],  String(body["_CMDLINE"]))       where body["_CMDLINE"] != nil
          - set(attributes["slice"],            String(body["_SYSTEMD_SLICE"])) where body["_SYSTEMD_SLICE"] != nil
          - set(attributes["runtime_scope"],    String(body["_RUNTIME_SCOPE"])) where body["_RUNTIME_SCOPE"] != nil
          #--- create low cardinality "job" identifier
          # user@xxx.service mnt-xxx.mount run-xxx.scope session-xxx.scope
          ## ^([a-zA-Z_]{3,20})   ([^\\-\\.\\@0-9]+).*
          - set(attributes["syslog_identifier_prefix"], ConvertCase(body["SYSLOG_IDENTIFIER"], "lower")) where body["SYSLOG_IDENTIFIER"] != nil
          - replace_pattern(attributes["syslog_identifier_prefix"], "^[^a-zA-Z]*([a-zA-Z]{3,25}).*", "$$1") where body["SYSLOG_IDENTIFIER"] != nil
          - set(attributes["unit_prefix"], ConvertCase(body["_SYSTEMD_UNIT"], "lower")) where body["_SYSTEMD_UNIT"] != nil
          - replace_pattern(attributes["unit_prefix"], "^[^a-zA-Z]*([a-zA-Z]{3,25}).*", "$$1") where body["_SYSTEMD_UNIT"] != nil
          - set(attributes["job"], attributes["syslog_identifier_prefix"])
          - set(attributes["job"], attributes["unit_prefix"]) where attributes["job"] == nil and attributes["unit_prefix"] != nil
          - delete_key(attributes, "syslog_identifier_prefix")
          - delete_key(attributes, "unit_prefix")
          #--- set timestamp from _SOURCE_REALTIME_TIMESTAMP if available
          - set(time, Time(Int(body["_SOURCE_REALTIME_TIMESTAMP"]), "microseconds")) where body["_SOURCE_REALTIME_TIMESTAMP"] != nil
          #--- set hostname and resource attributes
          - set(attributes["hostname"], String(body["_HOSTNAME"])) where body["_HOSTNAME"] != nil
          - set(resource.attributes["host.name"], String(body["_HOSTNAME"])) where body["_HOSTNAME"] != nil
          - set(resource.attributes["host.id"], String(body["_MACHINE_ID"])) where body["_MACHINE_ID"] != nil
          #--- set service.name according to OTEL specs (as resource attribute)
          # Priority: SYSLOG_IDENTIFIER > _SYSTEMD_UNIT > _COMM > fallback
          - set(resource.attributes["service.name"], String(body["SYSLOG_IDENTIFIER"])) where body["SYSLOG_IDENTIFIER"] != nil
          - set(resource.attributes["service.name"], String(body["_SYSTEMD_UNIT"])) where resource.attributes["service.name"] == nil and body["_SYSTEMD_UNIT"] != nil
          - set(resource.attributes["service.name"], String(body["_COMM"])) where resource.attributes["service.name"] == nil and body["_COMM"] != nil
          - set(resource.attributes["service.name"], "unknown_service") where resource.attributes["service.name"] == nil
          #--- also set service.version if available from systemd unit
          - set(resource.attributes["service.version"], "unknown") where resource.attributes["service.name"] != "unknown_service"
          #--- preserve important timing metadata as attributes
          - set(attributes["journald.boot_id"], String(body["_BOOT_ID"])) where body["_BOOT_ID"] != nil
          - set(attributes["journald.cursor"], String(body["__CURSOR"])) where body["__CURSOR"] != nil
          #--- remove journald metadata and make log body simple
          - set(attributes["message"], body["MESSAGE"]) where body["MESSAGE"] != nil
          #--- clean up: remove the large original_body to save storage (comment out if you need it for debugging)
          - delete_key(attributes, "original_body")
  groupbyattrs:
    keys:
      - service.name
      - host.name
      - receiver
      - job
```

**Comment by vidam-io** [comment]
@djaglowski @dmitryax
For now, I'm proceeding with [OTTL-based transformation](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7298#issuecomment-3187693464).

**Current Situation**
The OTEP for this has been approved, but no PR has been merged
The Journald receiver still passes raw JSON as the log Body, without mapping to proper attributes and severity
Most users in the comments are proceeding with transform processor as a workaround

**Broader Concerns**
This seems like a general pattern issue, not just Journald-specific. Similar challenges exist for other structured log sources (e.g., K8s Events).
I'm curious about the community's direction: Is relying on processors for semantic mapping (attributes, severity) the intended long-term approach? Or should receivers handle this more directly?

+) I've configured the SeverityText to be at the same level as the SeverityNumber.
